src/TMap.cpp

```diff
@@ -2649,6 +2653,550 @@ void TMap::setRoomNamesShown(bool shown)
+    mpProgressDialog = new QProgressDialog(tr("Import JSON map data to %1\n"
+                                              "Areas: %2 of: %3   Rooms: %4 of: %5   Labels: %6 of: %7...")
+                                                   .arg(mProfileName,
+                                                        QLatin1String("0"),
+                                                        QString::number(mProgressDialogAreasTotal),
+                                                        QLatin1String("0"),
+                                                        QString::number(mProgressDialogRoomsTotal),
+                                                        QLatin1String("0"),
+                                                        QString::number(mProgressDialogLabelsTotal)),
+                                           tr("Abort"),
+                                           0,
+                                           mProgressDialogRoomsTotal,
+                                           mpHost->mpConsole);
+    mpProgressDialog->setValue(0);
+    mpProgressDialog->setWindowModality(Qt::NonModal);
+    mpProgressDialog->setWindowTitle(tr("Map JSON import", "This is a title of a progress window."));
+    mpProgressDialog->setWindowIcon(QIcon(QStringLiteral(":/icons/mudlet_map_download.png")));
+    mpProgressDialog->setMinimumWidth(500);
+    mpProgressDialog->setAutoClose(false);
+    mpProgressDialog->setAutoReset(false);
+    mpProgressDialog->setMinimumDuration(1); // Normally waits for 4 seconds before showing
+    qApp->processEvents();
+
+    mDefaultAreaName = mapObj[scDEFAULT_AREA_NAME].toString();
+    mUnnamedAreaName = mapObj[scANONYMOUS_AREA_NAME].toString();
+    QString mapSymbolFontText = mapObj[scMAP_SYMBOL_FONT_DETAILS].toString();
+    float mapSymbolFontFudgeFactor = (qRound(mapObj[scMAP_SYMBOL_FUDGE_FACTOR].toDouble() * 1000.0)) / 1000;
+    bool isOnlyMapSymbolFontToBeUsed = mapObj[scONLY_USE_MAP_SYMBOL_FONT].toBool();
+    int playerRoomStyle = qRound(mapObj[scPLAYER_ROOM_STYLE].toDouble());
+    quint8 playerRoomOuterDiameterPercentage = qRound(mapObj[scPLAYER_OUTER_DIA_PERCENTAGE].toDouble());
+    quint8 playerRoomInnerDiameterPercentage = qRound(mapObj[scPLAYER_INNER_DIA_PERCENTAGE].toDouble());
+    QColor playerRoomOuterColor;
+    QColor playerRoomInnerColor;
+
+    if (mapObj.contains(scPLAYER_ROOM_COLORS) && mapObj.value(scPLAYER_ROOM_COLORS).isArray()) {
+        QJsonArray playerRoomColorArray = mapObj.value(scPLAYER_ROOM_COLORS).toArray();
+        if (playerRoomColorArray.size() == 2 && playerRoomColorArray.at(0).isObject() && playerRoomColorArray.at(1).isObject()) {
+            playerRoomOuterColor = readJsonColor(playerRoomColorArray.at(0).toObject());
+            playerRoomInnerColor = readJsonColor(playerRoomColorArray.at(1).toObject());
+        }
+    }
+
+    QMap<int, int> envColors;
+    if (mapObj.contains(scENV_TO_COLOR_MAPPING) && mapObj.value(scENV_TO_COLOR_MAPPING).isObject()) {
```

Comment:
This is just reading a simple key:value from the json but the code is not simple - could a range-for be used to simplify things here? Same with custom colors below, takes some effort to parse what it is doing - when it shouldn't.

Reply:
It is trying to be robust - first it checks for the thing to exist, then it checks that it is the right *sort* of thing (object vs array) - in other places where it is an array it also checks that it has a reasonable number of elements. I've seen other JSON stuff in a few places and it too is checking that something is not undefined before it tries to do anything doesn't it?

Reply:
Yeah but if we can get a ranged for, it won't even trigger if there's nothing there?